superset-frontend/packages/superset-ui-chart-controls/src/shared-controls/constants.tsx

```diff
@@ -58,26 +64,50 @@ export const xAxisMixin = {
   default: undefined,
 };
 
+export const getTemporalColumnsFromDatasouce = (
+  datasource: ValueOf<Pick<ControlPanelState, 'datasource'>>,
+) => {
```

Comment:
I see the use case for this helper function, but it feels slightly loaded. Also, would it make sense to place it in `superset-ui-core/utils`?

To make it slightly more globally applicable and with reference to [this comment](https://github.com/apache/superset/pull/21577#discussion_r979705093), would it be possible to call this `getColumnsFromDatasource` and add `datatypes?: GenericDataType[]` to the sig? Or alternatively call it `getColumnMetadata` and then return an object similar to 
```typescript
const rv: {
  defaultTemporalColumn?: string | null;
  columnsByDatatype: Record<GenericDataType, (ColumnMeta | QueryColumn)[]>;
} = ...
```
?

Reply:
Sure, let me refactor/move it to `superset-ui/core`.

Reply:
Hi @villebro, I found after trying to refactor that this part might require more refactoring working in SQLLab result set. Because the result set in SQLLab lack of `GenericDataType`. So, I think I move this function to the `superset-control/utils` first. The more refactor work should open a new PR for coordinating data results between a **query metadata** and the **superset datasource metadata**.

<img width="1348" alt="image" src="https://user-images.githubusercontent.com/2016594/192488857-bd03e9bf-f638-4e97-887d-04d1969966ac.png">

Reply:
Yeah, I understand. I was wondering why there was that check for `results`, but it makes sense now.